pageserver/src/layered_repository.rs

```diff
@@ -168,7 +168,19 @@ impl Repository for LayeredRepository {
 
     /// Branch a timeline
     fn branch_timeline(&self, src: ZTimelineId, dst: ZTimelineId, start_lsn: Lsn) -> Result<()> {
-        let src_timeline = self.get_timeline(src)?;
+        // We need to hold this lock because gc scans directory to learn about timelines,
+        // so race condition is possible, we create new timeline and gc can remove data that
+        // might be referenced by newly created timeline.
```

Comment:
This is a bit hard to read, maybe:
```suggestion
        // We need to hold this lock to prevent GC from starting at the same time. GC scans the directory to learn 
        // about timelines, so otherwise a race condition is possible, where we create new timeline and GC
        // concurrently removes data that is needed by the new timeline.
```